### PR TITLE
If running under k8ssandra, wait for the CassandraDatacenter

### PR DESCRIPTION
**Pull request by burmanm**
Wait for DC to be ready before deploying schema job. Fixes https://github.com/k8ssandra/k8ssandra/issues/45

However, this only works when running under k8ssandra and is highly dependent on the helm chart's structure. For those reasons, if such structure is not found, this version will simply not wait and fallback to the previous behavior. 

I'm not super satisfied with that approach, however given the original ticket - we don't necessarily have the CassandraDatacenter is other deployment scenarios so depending on it while keeping the changes in this repo is somewhat necessary.

### CODE REVIEW

File: api/v1alpha1/reaper_types.go
Comment (jsanda) [inline api/v1alpha1/reaper_types.go:79]: When I originally wrote a lot of this, there was no integration or even planned integration with cass-operator. Now though we are tightly integrating with cass-operator which simplifies some things. In the `CassandraBackend` struct we can add a `CassandraDatacenter` property along these lines:

```go
type CassandraDatacenterRef struct {
        Name string `json:"name"`

        // If empty we could default the Reaper namespace.
        Namespace string `json:"namespace,omitempty"`
}

type CassandraBackend struct {
       CassandraDatacenter CassandraDatacenterRef `json:"cassandraDatacenter"`

       // otrher fields....
}
```

The `ClusterName` and `CassandraService` fields in `CassandraBackend` could then go away. This will eliminate the need for the `DatacenterName()` function.

When you update type definitions, be sure to run `make manifests` to regenerate updated CRDs.

File: pkg/reconcile/reconcilers.go
Comment (jsanda) [inline pkg/reconcile/reconcilers.go:163]: The pattern I have been following when there is an error is to log the error, and then requeue the request and return the error.

File: pkg/reconcile/reconcilers.go
Comment (jsanda) [inline pkg/reconcile/reconcilers.go:196]: You should not need to check for a not registered error. We register the CassandraDatacenter schema at start up [here](https://github.com/thelastpickle/reaper-operator/blob/master/main.go#L49).

File: pkg/reconcile/reconcilers.go
Comment (jsanda) [inline pkg/reconcile/reconcilers.go:208]: I think it is generally best to avoid potentially long blocking operations during reconciliation. You want a workflow along these lines:

* Fetch the CassandraDatacenter
* If fetch fails
    * Log the error
    * Requeue the request
* Check the status
* If the CassandraDatacenter is not ready
    * Log a message
    * Request the request
* If the CassandraDatacenter is ready
    * Log a message
    * Continue with reconciliation 

And if we really wanted to get fancy, we could calculate the requeue delay bases on the C* cluster size and the number of up C* nodes. We can  hold off on that for now :)

File: pkg/reconcile/reconcilers.go
Comment (jsanda) [inline pkg/reconcile/reconcilers.go:161]: See my comment below about `checkForCassandraDatacenterReadiness`. Basically the check should not block and should fail fast.

**Review by jsanda — CHANGES_REQUESTED** [review]
Let me know if you have any questions about my comments. Looking good so far!

File: pkg/reconcile/reconcilers.go
Comment (burmanm) [reply pkg/reconcile/reconcilers.go:196]: Without the check, the tests will fail.

Comment (burmanm) [reply pkg/reconcile/reconcilers.go:196]: Although I guess these are not missing on purpose from the suite_test.go .. I'll try fixing it.

Comment (jsanda) [reply pkg/reconcile/reconcilers.go:196]: Yeah, the CassandraDatacenter scheme  needs to be added in the integration test. Sorry, I was only thinking before about the e2e test.

**Comment by burmanm** [comment]
With latest changes (and modifications to k8ssandra charts):

```
[michael@nina datastax]$ kubectl get pods
NAME                                                   READY   STATUS      RESTARTS   AGE
cass-operator-86d4dc45cd-76l7v                         1/1     Running     0          3m17s
demo-kube-prometheus-stack-operator-7cfdf8c7cb-7c25n   1/1     Running     0          3m17s
grafana-deployment-695d6dbcd7-5qvv6                    1/1     Running     0          119s
k8sc-grafana-operator-k8ssandra-5b66978766-td595       1/1     Running     0          3m
k8sc-reaper-k8ssandra-6f48575445-k5kbx                 0/1     Running     0          55s
k8sc-reaper-k8ssandra-schema-cpmjh                     0/1     Completed   0          65s
k8sc-reaper-operator-k8ssandra-7c7dd9bb4f-9dck9        1/1     Running     0          3m
k8ssandra-dc1-default-sts-0                            2/2     Running     0          2m57s
prometheus-k8sc-prometheus-k8ssandra-0                 2/2     Running     1          3m
[michael@nina datastax]$
```

File: go.mod
Comment (jsanda) [inline go.mod:17]: Is this all that was needed to be able to register the CassandraDatacenter scheme in the integration tests?

File: pkg/reconcile/reconcilers.go
Comment (jsanda) [inline pkg/reconcile/reconcilers.go:205]: We actually do not need to check for the `reaper.cassandra-reaper.io/instance` annotation here. The purpose of that annotation is to tell reaper-operator that the CassandraDatacenter should be managed by Reaper for repairs. The backend storage does not have to managed for repairs.

File: pkg/reconcile/reconcilers.go
Comment (jsanda) [inline pkg/reconcile/reconcilers.go:228]: Shouldn't we also return the error returned from the `r.Client.Get(ctx, cassdcKey, cassdc)` call?

**Review by jsanda — CHANGES_REQUESTED** [review]
There are couple small things I noticed. Other than that, things look really good!

File: pkg/reconcile/reconcilers.go
Comment (burmanm) [reply pkg/reconcile/reconcilers.go:228]: Problem is that it doesn't work in the tests (it will always throw an error).

**Comment by jsanda** [comment]
@burmanm I merged #8. Can you please rebase off of master? This should unblock you on the integration test error.

**Review by jsanda — CHANGES_REQUESTED** [review]
I have transferred the project to the k8ssandra GitHub org. Can you please rebase? 

Sorry if it causes any conflicts.

File: controllers/reaper_controller_test.go
Comment (jsanda) [inline controllers/reaper_controller_test.go:61]: Do we need this `k8sClient.Update()` call here?

File: pkg/reconcile/reconcilers.go
Comment (jsanda) [inline pkg/reconcile/reconcilers.go:227]: Should the `Printf` call be removed?

**Review by jsanda — CHANGES_REQUESTED** [review]
Just a couple minor things I saw, and then I think this is good to go.

**Review by jsanda — APPROVED** [review]

**Pull request merged**
